### PR TITLE
[LLVM][Coroutines] Create `.noalloc` variant of switch ABI coroutine ramp functions during CoroSplit

### PR DESCRIPTION
**Pull request by yuxuanchen1997**
This patch is episode two of the coroutine HALO improvement project published on discourse: https://discourse.llvm.org/t/language-extension-for-better-more-deterministic-halo-for-c-coroutines/80044

Previously CoroElide depends on inlining, and its analysis does not work very well with code generated by the C++ frontend due the existence of many customization points. There has been issue reported to upstream how ineffective the original CoroElide was in real world applications. 

For C++ users, this set of patches aim to fix this problem by providing library authors and users deterministic HALO behaviour for some well-behaved coroutine `Task` types. The stack begins with a library side attribute on the `Task` class that guarantees no unstructured concurrency when coroutines are awaited directly with `co_await`ed as a prvalue. This attribute on Task types gives us lifetime guarantees and makes C++ FE capable to telling the ME which coroutine calls are elidable. We convey such information from FE through the attribute `coro_elide_safe`.

This patch modifies CoroSplit to create a variant of the coroutine ramp function that 1) does not use heap allocated frame, instead take an additional parameter as the pointer to the frame. Such parameter is attributed with `dereferenceble` and `align` to convey size and align requirements for the frame. 2) always stores cleanup instead of destroy address for `coro.destroy()` actions. 

In a later patch, we will have a new pass that runs right after CoroSplit to find usages of the callee coroutine attributed `coro_elide_safe` in presplit coroutine callers, allocates the frame on its "stack", transform those usages to call the `noalloc` ramp function variant. 

(note I put quotes on the word "stack" here, because for presplit coroutine, any alloca will be spilled into the frame when it's being split)

The C++ Frontend attribute implementation that works with this change can be found at https://github.com/llvm/llvm-project/pull/99282
The pass that makes use of the new `noalloc` split can be found at https://github.com/llvm/llvm-project/pull/99285

### CODE REVIEW

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (vogelsgesang) [inline llvm/lib/Transforms/Coroutines/CoroSplit.cpp:2053]: Additionally, we should only create this `noalloc` variant, if the coroutines promise type itself is marked as "elidable".

**Example**

```
struct [[coro_must_elide]] ElidedTask { ... }
struct NonElidedTask { ... }

ElidedTask foo();
NonElidedTask bar();

NonElidedTask foobar() {
    co_await foo();
    co_await bar();
}
```

**Current behavior**: We create `foo.noalloc`, `bar.noalloc` and `foobar.alloc`.

`foo.noalloc` gets actually called. However, `bar.noalloc` and `foobar.alloc` are dead code. The `.noalloc` variants will never be used, as the `NonElidedTask` is not marked with `[[coro_must_elide]]`.

I think we should avoid generating this dead code

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (vogelsgesang) [inline llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1480]: I wonder if the function should be marked as `linkage = internal`. To my understanding, the `[[coro_must_elide]]` does not work across translation units, yet. As such, there won't be any cross-TU calls to the `.noalloc` function

**Review by vogelsgesang — APPROVED** [review]
LGTM on a high level.

But I am not experienced in this code area, so you might want to wait for another review

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (yuxuanchen1997) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1480]: Does ThinLTO count as Cross-TU?

Comment (vogelsgesang) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1480]: Intuitively, I would think so, yes. But I never looked into ThinLTO before, so I don't know for sure.

Does #99285 enable `CoroAnnotationElide` already for ThinLTO?

Comment (apolloww) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1480]: ThinLTO can import global values (variables and functions) from another modules. There would be a follow-up patch updating the pipeline so that coroutine passes are moved from pre-link to post-link. A previous attempt #90310 was blocked due to some conflict with Asan. I'll have a new version later.

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (yuxuanchen1997) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:2053]: We currently don't propagate the attribute from the CXXRecordDecl to the LLVM type. We can potentially add another attribute at the coroutine function level if that sounds good to you.

Comment (vogelsgesang) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:2053]: Yes, I would be in favor of adding a second attribute. What do you & others think?
CC @ChuanqiXu9

Comment (yuxuanchen1997) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:2053]: > Yes, I would be in favor of adding a second attribute. What do you & others think? CC @ChuanqiXu9

I can put up another PR for that. Wdyt?

Comment (vogelsgesang) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:2051]: fine by me. Doesn't need to be addressed as part of this commit, can also be done in a follow-up commit.
Maybe add a `FIXME` to the source code, such that other people reading the source code in the meantime are aware of this future work?

Comment (yuxuanchen1997) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:2051]: For sure!

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (ChuanqiXu9) [inline llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1465]: We need comment and an example for this function

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (ChuanqiXu9) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1478]: I think this should be marked as internal like .resume and .destroy function does. While LTO can import them, I don't feel we can benefit a lot from that.

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (ChuanqiXu9) [inline llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1467]: If I am not mistaken, the identation looks not good here.

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (ChuanqiXu9) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:2051]: Agreed.

File: llvm/test/Transforms/Coroutines/coro-split-00.ll
Comment (ChuanqiXu9) [inline llvm/test/Transforms/Coroutines/coro-split-00.ll:66]: I think it worth to list the arguments explicitly.

File: llvm/test/Transforms/Coroutines/ArgAddr.ll
Comment (ChuanqiXu9) [inline llvm/test/Transforms/Coroutines/ArgAddr.ll:8]: Are these changes to `@llvm.coro.id` related to this patch really?

**Review by ChuanqiXu9 — COMMENTED** [review]
We need to update coroutines.rst for such changes.

**Comment by ChuanqiXu9** [comment]
After I took a quick look at https://github.com/llvm/llvm-project/pull/99285, I feel this is not what I thought when I heard the idea. Correct me if I misunderstand it.

The problems are:
1. It looks like all the `.noalloc` variant are emitted all the time. This is absolutely not good. It emits duplicated code (especially currently its linkage is not internal). What I had in mind is to create `.noalloc` variant on need or lazily.
2. The original ramp function and the `.noalloc` variant shares a lot of codes if I am not mistaken. Then it is pretty bad for the code size. What I had in mind is, after we generate the `.noalloc` variant, we will rewrite the ramp function too and the ramp function will call the `.noalloc` function after it allocates the coroutine frame.

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (yuxuanchen1997) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1467]: Didn't get that?

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (yuxuanchen1997) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1478]: This pass is now configured to not run at all during LTO pre-link. So internal linkage SGTM.

File: llvm/test/Transforms/Coroutines/ArgAddr.ll
Comment (yuxuanchen1997) [reply llvm/test/Transforms/Coroutines/ArgAddr.ll:8]: Ah, the `noalloc` variant was appended to `f.resumers`, causing the name to be changed to `f.resumers.1`.

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (ChuanqiXu9) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1467]: Never mind. I made a silly mistake.

File: llvm/test/Transforms/Coroutines/ArgAddr.ll
Comment (ChuanqiXu9) [reply llvm/test/Transforms/Coroutines/ArgAddr.ll:8]: I feel this is something need to be fixed. It looks odd to treat the  noalloc variant as resumers.

Comment (yuxuanchen1997) [reply llvm/test/Transforms/Coroutines/ArgAddr.ll:8]: Other than the legacy `CoroElide`, the main purpose of this variable is to keep connection on the graph between the original copy and the `.resume`, `.destroy`, `.cleanup`, `.noalloc` splits. I would suggest just changing the name to read "f.splits" instead of `f.resumers`. It's a requirement for CGSCC call graph updating.

**Comment by yuxuanchen1997** [comment]
> 1. It looks like all the `.noalloc` variant are emitted all the time. This is absolutely not good. It emits duplicated code (especially currently its linkage is not internal). What I had in mind is to create `.noalloc` variant on need or lazily.

In my experience, ramp functions are generally very small (single basic block after all the stripping, etc.). However, this might not be the case for other coroutines and I understand that we need to have a mechanism in place to generate such variant only for cases where it's necessary. 

> What I had in mind is, after we generate the `.noalloc` variant, we will rewrite the ramp function too and the ramp function will call the `.noalloc` function after it allocates the coroutine frame.

I originally also had this idea. One would need to properly substitute the cleanup function with the destroy one and legacy CoroElide may no longer work in this case, which is insignificant of course considering what this patch is doing. I don't know whether it's worth additional complexity. 

If we decide that it's worth it, we can implement it in a separate patch so that we don't risk changing too much at once.

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (yuxuanchen1997) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:2062]: This turned out to be easy. I am addressing this with the next push for this patch.

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (yuxuanchen1997) [reply llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1471]: This is done. Thanks.

**Comment by yuxuanchen1997** [comment]
@ChuanqiXu9 I have changed this patch to only conditionally create the `.noalloc` variant based on an attribute (which is controlled by FE). Let me know if this is good to go.

File: llvm/docs/Coroutines.rst
Comment (vogelsgesang) [inline llvm/docs/Coroutines.rst:2062]: I think the name `coro_must_elide` is a misnormer. "must elide" sounds as if it would be a compilation error if elision fails. However, this is not the case

Comment (yuxuanchen1997) [reply llvm/docs/Coroutines.rst:2062]: What about `coro_elide_safe`?

Comment (vogelsgesang) [reply llvm/docs/Coroutines.rst:2062]: love it!

Comment (yuxuanchen1997) [reply llvm/docs/Coroutines.rst]: Changed. Closing this one.

File: llvm/docs/Coroutines.rst
Comment (ChuanqiXu9) [inline llvm/docs/Coroutines.rst:2068]: It will be better to explain and describe the `f.noalloc` ramp function  in this document. And it will be better to have some  example codes for it and compare it with the normal ramp functions.

File: llvm/lib/Transforms/Coroutines/CoroInternal.h
Comment (ChuanqiXu9) [inline llvm/lib/Transforms/Coroutines/CoroInternal.h:29]: Let's add some comments for this since I can't guess its job by its name.

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (ChuanqiXu9) [inline llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1502]: nit: it looks better to use `FrameIdx` below instead of using the induction variable across code sections.

File: llvm/lib/Transforms/Coroutines/CoroSplit.cpp
Comment (ChuanqiXu9) [inline llvm/lib/Transforms/Coroutines/CoroSplit.cpp:1474]: nit: I feel better with an assertion here that the ABI is switch ABI.

**Review by ChuanqiXu9 — COMMENTED** [review]
The patch looks good to me except the thing I mentioned in https://github.com/llvm/llvm-project/pull/99282#pullrequestreview-2265588601

File: llvm/docs/Coroutines.rst
Comment (yuxuanchen1997) [reply llvm/docs/Coroutines.rst:2068]: This attribute is deleted while addressing your feedback in https://github.com/llvm/llvm-project/pull/99282#pullrequestreview-2265588601

I can add a clarification in the documentation for coro_safe_elide.

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM

**Pull request merged**
